apps/champions/lib/champions/battle/simulator.ex

```diff
@@ -63,187 +64,459 @@ defmodule Champions.Battle.Simulator do
-         _caster,
-         target
-       ),
-       do: target[String.to_atom(effect.stat_affected)] + effect.amount
+  # Return whether an effect with a ChanceToApply component hits.
+  # Later on, this might also handle similar mechanics like the target's dodge chance.
+  defp effect_hits?(effect) do
+    chance_to_apply_component =
+      Enum.find(effect.components, fn comp -> comp["type"] == "ChanceToApply" end)
 
-  defp calculate_value(
-         %{amount_format: "multiplicative", stat_based_on: nil} = effect,
-         _caster,
-         target
-       ),
-       do: target[String.to_atom(effect.stat_affected)] * effect.amount
+    case chance_to_apply_component do
+      nil ->
+        true
 
-  defp calculate_value(
-         %{amount_format: "additive", stat_based_on: stat_based_on} = effect,
-         caster,
-         target
-       ),
-       do:
-         target[String.to_atom(effect.stat_affected)] +
-           floor(effect.amount * caster[String.to_atom(stat_based_on)] / 100)
+      chance_to_apply_component ->
+        chance_to_apply_component["chance"] >= :rand.uniform()
+    end
+  end
 
+  # Apply an effect to its target. Returns the new state of the target.
+  # For now this applies the executions on the spot.
+  # Later on, it will "cast" them as we do with skills and effects to account for execution delays.
+  # Returns the new state of the target.
+  defp apply_effect(effect, target, caster, current_step_number) do
+    target_after_modifiers =
+      Enum.reduce(effect.modifiers, target, fn modifier, target ->
+        # If it's permanent, we set its duration to -1
+        new_modifier =
+          modifier
+          |> Map.put(:remaining_steps, Map.get(effect.type, "duration", -1))
+          |> Map.put(:step_applied_at, current_step_number)
+
+        case modifier.modifier_operation do
+          "Add" ->
+            put_in(target, [:modifiers, :additives], [new_modifier | target.modifiers.additives])
+
+          "Multiply" ->
+            put_in(target, [:modifiers, :multiplicatives], [new_modifier | target.modifiers.multiplicatives])
+
+          "Override" ->
+            put_in(target, [:modifiers, :overrides], [new_modifier | target.modifiers.overrides])
+        end
+      end)
+
+    target_after_executions =
+      Enum.reduce(effect.executions, target_after_modifiers, fn execution, target_acc ->
+        process_execution(execution, target_acc, caster)
+      end)
+
+    target_after_executions
+  end
+
+  # Apply a DealDamage execution to its target. Returns the new state of the target.
+  defp process_execution(
+         %{
+           "type" => "DealDamage",
+           "attack_ratio" => attack_ratio,
+           "energy_recharge" => energy_recharge,
+           # TODO
+           "delay" => _delay
+         },
+         target,
+         caster
+       ) do
+    damage = max(floor(attack_ratio * calculate_unit_stat(caster, :attack)), 0)
```

Comment:
this approach to the modifiers causes to calculate the attack every time we wan't to deal damage , why not modifying the damage when we add or remove the effects and then simply use it? let me know if you think this simplify the logic or just makes it harder

something like this pseudo code
apply_effect -> 
 added amount = calculate amount
 unit.attribute += added amount
  map.put(modifier, :added_amount, added_amount)
add modifier to unit
remove_effect -> 
 unit.attribute -= effect.added_amount
 remove modifier from unit

Reply:
The reason is that modifiers don't necessarily expire in the same order they were applied, so this could happen:

1. Unit has 100 base_damage
2. A modifier that expires in a short time reduces his damage by 20 (additive). Now its 80.
3. A modifier that expires in a long time reduces his damage in half (multiplicative). Now its 40.
4. The additive expires, we restore the 20. Now we have 60.
5. The multiplicative expires, we restore the halving of the damage. Now its 120.

The unit now has more damage than what it started with. This is bad! Even if we found a workaround for restoring, the result for the stat during modifiers would depend on the order they were applied, which is also probably bad

Reply:
I think this problem of modifying some attibute from a player is very common since i've seen it in curse of mirra before too 😂  ok it has sense now! thanks